### ISSUE TITLE
Combine @EnableAuthorizationServer and @EnableOAuth2Sso

### ISSUE THREAD

**Issue by Chumper** [spring-projects/spring-boot#3965]
Hey, want to create a service that will act as a client for an oauth2 server but should also serve as authorization server for other services.

I am using the JWT solution with public/private keys btw.

My setup looks like this:

I have one central authentication server, that is build with spring social.
It will issue JWT tokens to its client, that are signed with the private key of this server.

The client will fetch the public key at runtime:
```yaml 
security:
  oauth2:
    resource:
      jwt:
          key-uri: "http://localhost:8081/oauth/token_key"
```
and can verify the token.

It should also act as a authentication server for services under this server, so i added:
```java
@EnableAuthorizationServer
class AuthorizationServerConfig extends AuthorizationServerConfigurerAdapter {
...
@Bean
    public JwtAccessTokenConverter tokenEnhancer() {
        //set private and public key here
        ...
    }
}
```

Unfortunately this will override the fetched key from the authorization server, so a token issued by the first server is not valid anymore, because the public key was overriden.

How can i solve this?

My dirty workaround atm is to extend `JwtAccessTokenConverter` and allow multiple verifier keys to be present.
As long as one key validates the token it is accepted.

Maybe this is the wrong project but as oauth2 is a subproject of boot from 1.3 i thought i post it here

**Comment by dsyer** [comment]
I'm confused really. If there is "one central authentication server, that is build with spring social" then why do we need `@EnableAuthorizationServer`? There are 2 authorization servers? Why? Maybe you only need `@EnableResourceServer` on the client (even then I'm not sure why yet, since it is a client, not a resource - although I suppose it could be both).

There is no "subproject" of boot which is relevant here really either, as far as I can tell, but let's try and bottom out the actual requirement before we move the question elsewhere.

**Comment by Chumper** [comment]
Hey, i will try to explain the setup i want to have:

I have a central authentication server, where all my user data is stored, based on `Spring Social` so you can log in with facebook, google etc.

However, a `User` can have many `Accounts` on my `Service`, so i wanted to add a second authentication server, that will convert the `User` token into a `Service` token by choosing which `Account` the user will use on the `Service`. 

Therefore it needs to accept central tokens and convert them to service tokens. So a flow would be like:
* User wants to use the service, but will be redirected the the central - service token converter (internal oauth2 flow start)
* User will be redirected to central authentication (external oauth2 flow start)
* User logs in (facebook, google etc)
* CA will generate a central token for the user
* Will get redirected to the central - service token converter (external oauth2 flow ends)
* Central - service token converter will generate a service token for the user
* Will get redirected to service (internal flow ends)

So i have three servers involved, where one is client and authorization server in one.
Hope this makes my requirement clear or let me know if this idea is completely stupid.

**Comment by dsyer** [comment]
Thanks for the clarification. I guess I don't really see the need for 2 internal authorization servers (3 including the external provider). Can't you just use the one (plus the external)? If not then I don't understand what the problem is anyway. Something to do with JWT tokens - both your internal auth servers issue JWTs signed with different keys? Even then I don't understand why that's a problem, expect that it sounds very complicated and I'm not sure why it is necessary.

**Comment by Chumper** [comment]
You are correct, in the worst case i have 3 nested oauth2 flows. If I can keep it simple i would love to do it, but for my use case that is the minimal setup i can currently imagine.

My problem with just one authorization server is that with that setup one `User` equals to one `Account` on a `Service` and i can not have multiple `Accounts` for a `User`.

Lets say i have game `Service`. I would not be able to offer multi `Accounts` on this game `Service`, as each `User` has exactly one `Account`.

In order to offer multiple `Accounts` i need a second authorization server that translates between `Users` and `Accounts` and offer some kind of account switching.

The other solution would be to add the `Account` information to the JWT so the `Service` can authenticate the user.
But what would i do if the `User` has no `Account` on this `Service` so i cant include any information into the JWT?

As said above, if there is a simple way to do this i would love to hear it, maybe i am just missing some key elements.

**Comment by dsyer** [comment]
I don't really know what an `Account` or a `Service` is in your domain so it's hard to really help much. I can't really believe it needs to be this complicated though, if you control the code in both "auth" servers.

**Comment by Chumper** [comment]
Hey Dave, sorry that i am not clear about my use case. Maybe you are right and i don not need this many authorzation servers, so i am trying to reduce this to just one.
In fact I maybe found a solution to enhance the central token with my additional information and use it in the service. I will give it a try.

Anyway here is my definition:

`User`: the person using the `Services` e.g. **Bob**
`Service`: a group of servers representing a single application. e.g. a game backend, website backend etc
`Account`: Information that uniquely identify a `User` on a `Service`. e.g. **Bob** has one game `Account` and two website `Accounts`

Basically it comes down to multi accounts usage with just a single JWT token and how that can be realised. One user, many accounts, while the accounts can be on different applications.
Including all account information in the JWT will be bad because then the token will just grow.
So i need a way to include the right account information into the token.
